src/darcy.erl

```diff
@@ -421,6 +482,185 @@ process_result_set(#{ <<"Items">> := Items, <<"Count">> := C }) ->
+        {Results, []} -> {ok, Results};
+        {Partial, Errors} -> {error, {Errors, Partial}};
+        Other -> Other
+    after Timeout ->
+        {error, scan_timeout}
+    end.
+
+%% @doc Scan a table in parallel using the given expression.
+%%
+%% DynamoDB supports parallel scans using a partitioning technique described in the Developer Guide.
+%%
+%% https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Scan.html
+%%
+%% This function spawns a coordinator process which in turn spawns a set of
+%% workers, one per segment count which executes the `scan_all/4' function.
+%%
+%% <B>N.B.</B>: This operation can consume a lot of read capacity. It is a good
+%% idea to limit the number of segments used in a scan operation.
+%%
+%% The coordinator that is spawned here is <B>linked</B> to the caller. If you
+%% want more robust error handling, you should trap exit messages by the
+%% caller's process.
+%%
+%% Since this function returns the "raw" results, you will have to
+%% handle them appropriately within your own receive block.
+-spec scan_parallel(
+        Client :: darcy_request:aws_client(),
+        TableName :: binary(),
+        Expr :: map(),
+        Fun :: function(),
+        Timeout :: pos_integer(),
+        SegmentCount :: pos_integer(),
+        ReplyPid :: pid()) -> scan_timeout |
+                              { Results :: [ term() ], Errors :: [ term() ] }.
+scan_parallel(Client, TableName, Expr, Fun, _Timeout, SegmentCount, ReplyPid) ->
+    Reqs = [ make_scan_request([Expr, table_name(TableName), make_segments(N, SegmentCount)]) ||
+            N <- lists:seq(0, SegmentCount - 1) ],
+
+    spawn_link(fun() -> start_coordinator(Client, Fun, Reqs, ReplyPid) end).
+
+start_coordinator(Client, Fun, Reqs, Reply) ->
+    Pids = [ spawn_link(fun() -> worker_scan_all(Client, R, Fun, self()) end) || R <- Reqs ],
+    coordinator_loop(Pids, Reply, [], []).
+
+worker_scan_all(Client, Request, Fun, Coordinator) ->
+     Results = do_scan_all(Client, Request, execute_scan(Client, Request), Fun, []),
+     Coordinator ! {self(), Results}.
+
+coordinator_loop([], Reply, Acc, Errors) -> Reply ! {lists:reverse(lists:flatten(Acc)), lists:reverse(Errors)};
+coordinator_loop(Pids, Reply, Acc, Errors) ->
+    receive
+        {Pid, {ok, Result}} ->
+            coordinator_loop( Pids -- [Pid], Reply, [ Result | Acc ], Errors);
+        {Pid, {error, Error, Partial}} ->
+            coordinator_loop( Pids -- [Pid], Reply, [ Partial | Acc ], [ Error | Errors ]);
+        Other ->
+            %% fall through for rando messages
+            error_logger:warning_msg("Received unexpected message: ~p", [Other]),
+            coordinator_loop(Pids, Reply, Acc, Errors)
+    after ?BIG_TIMEOUT * 5 ->
+        error_logger:info_msg("The scan has timed out."
+                              "Worker pids: ~p, reply pid: ~p, accumulator: ~p, errors: ~p",
+                              [Pids, Reply, Acc, Errors]),
+        Reply ! scan_timeout
```

Comment:
should it be {error, scan_timeout} ?

Reply:
hmm maybe. for now Im going to leave it as-is